src/main/java/me/itzg/tsdbcassandra/services/DataTablesStatements.java

```diff
@@ -0,0 +1,80 @@
+package me.itzg.tsdbcassandra.services;
+
+import java.time.Duration;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Component;
+
+/**
+ * Provides a caching layer for insert and query statements that execute against the configuration-driven
+ * data tables schema.
+ */
+@Component
+public class DataTablesStatements {
+
+  public static final String TENANT = "tenant";
+  public static final String SERIES_SET = "series_set";
+  public static final String AGGREGATOR = "aggregator";
+  public static final String TIMESTAMP = "ts";
+  public static final String VALUE = "value";
+
+  public String tableNameRaw() {
+    return "data_raw";
+  }
+
+  public String tableNameDownsampled(Duration granularity) {
+    return "data_" + granularity.toString().toLowerCase();
+  }
+
+  /**
+   * @return an INSERT CQL statement with placeholders
+   * tenant, seriesSet, timestamp, value
+   */
+  @Cacheable("insertRaw")
+  public String insertRaw() {
+    return "INSERT INTO " + tableNameRaw()
+        + " (" + String.join(",", TENANT, SERIES_SET, TIMESTAMP, VALUE) + ")"
+        + " VALUES (?, ?, ?, ?)";
+  }
+
+  /**
+   * @return an INSERT CQL statement with placeholders
+   * tenant, seriesSet, aggregator, timestamp, value
+   */
+  @Cacheable("insertDownsampled")
```

Comment:
you are just caching strings here, right?  is the string building that expensive?  at first, i thought this was a mechanism for avoiding redundant writes to the metadata tables.

Reply:
Yes, that's all the caching is helping here. More importantly than the CPU cost of string building was wanting to minimize garbage collection of the short-lived strings during the ingest path. It's pre-optimizing, but seemed worth it for ingest and downsample call paths. I add the query string building since I needed at least the one during downsampling.